### ISSUE TITLE
[cssom-view] Consider making scroll methods return promises

### ISSUE THREAD

**Issue by mkay581** [w3c/csswg-drafts#1562]
[Smooth scrolling](https://drafts.csswg.org/cssom-view/#smooth-scrolling) has been introduced into the spec where the scroll happens over a period of time, which is great. However, I think it would be even more useful if scroll method returned a promise so the developer could wait until the scroll finishes before moving on to next set of tasks. 

Obviously `scroll()`ing on an element where `scroll-behavior` option set to `auto` makes scroll instant and is therefore not inherently asynchronous. But I think, to keep the API consistent, the scroll method should return a promise every time it is used.

So if you had the following HTML...

```html
<div id="scroll-container" style="scroll-behavior: smooth;">
   <!-- other content here --> 
</div>
```

you can do things like this:

```js
element.scroll(0, 400).then(() => {
   // ...scroll has finished
});
```

**Comment by mkay581** [comment]
FWIW, [here](https://discourse.wicg.io/t/proposal-element-scroll-with-ease-options/1720/6?u=mkay581) is a link to the relevant discussion where others have also +1'd this feature.

**Comment by dartess** [comment]
in 2018 it is still very much needed, only now we will use it this way:

```
let element = document.getElementById('scroll-container');
await element.scroll(0, 400);
// ...scroll has finished
```

use case: set focus after scroll

**Comment by Slapbox** [comment]
Is there an ETA for this being spec?

**Comment by patchthecode** [comment]
still waiting...

**Comment by webdevelopers-eu** [comment]
Unlike TabAtkins' opinion in the IRC chat, I strongly support the idea of implementing Promise rejection if a scroll is interrupted.

As I mentioned in #3744 , the intended effect of `scrollIntoView` is to bring the specified element into view. If this does not happen for any reason, the promise should be rejected, as the action was not fulfilled. There are specific scenarios where this should occur, such as:

* The element is removed or stops being rendered before it reaches the visible part of the viewport.
* The smooth, time-dependent scrolling is interrupted programmatically or by user action.

Simply put, the goal of `scrollIntoView` is to bring an element into view. If this is achieved, the promise should be fulfilled; if not, the promise should be rejected. This aligns with the natural and expected behavior that any programmer would anticipate when triggering an action intended to produce a specific result.

**Comment by markcellus** [comment]
> If this is achieved, the promise should be fulfilled; if not, the promise should be rejected. 

This is a valid point @webdevelopers-eu. But I think throwing an error may be unexpected for these cases and developers are probably likely to miss handling it. Is it possible to just abort/cancel the promise instead? That would make it so that an app handles this gracefully (by default) instead of abruptly throwing an error, which could potentially cause an entire website from working if it's a SPA (or javascript-heavy).

**Comment by webdevelopers-eu** [comment]
In that case, I can't see any other situation where a Promise rejection would occur, which essentially reduces the Promise to just being a callback. This undermines its purpose, as Promises are intended to handle success and failure cases, not just one. If we’re only using it to signal success and merely aborting in failure scenarios, we strip away the robustness of the Promise pattern and lose the benefit of built-in error handling.

**Comment by tabatkins** [comment]
In an async function, the promise is rejected *only* when the function throws an error. All other cases result in a fulfilled promise. This pattern is generally followed by other DOM functions that return promises; rejection is reserved for errors where we'd throw in a synchronous function.

In this case, we're just using the promise for its timing aspect.

**Comment by flackr** [comment]
> In this case, we're just using the promise for its timing aspect.

Additionally, the promise can be passed an argument which includes the completion status of the scroll so that developers can choose to have different behavior depending on whether the scroll was completed or interrupted.

**Comment by mustaqahmed** [comment]
While implementing this, we found two Blink tests that are affected by a minor error handling change caused by this proposal here.  I don't expect this corner case to be a compat problem, and wanted to check what other folks here think:

A JS parsing error caused by a malformed `scroll*()` call would no longer throw synchronously; instead it would return a Promise that would throw asynchronously.  For example, in Blink the `shouldThrow('window.scrollTo(x)')` call [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/fast/dom/Window/window-scroll-arguments.html;drc=77578ccb4082ae20a9326d9e673225f1189ebb63;l=41) fails because the `eval()` is seeing a `Promise` object instead of an exception.

Anyone concerned?

**Comment by tabatkins** [comment]
I'm not too concerned, since that's a mistake in the first place and would be fixed by the author. Only places explicitly catching a thrown error and doing something meaningful with it would notice.

**Comment by jeremenichelli** [comment]
I am super mega late to the party since I saw this on social media (sorry for the noise). But I cam to leave a comment that pretty much matches this one: https://github.com/w3c/csswg-drafts/issues/1562#issuecomment-2997993316

I wonder why the pattern used for the Web Animation API wasn't considered:

```js
const elem = document.getElementById('test');
await elem.scroll(0, 400).finished;
```

Besides matching APIs we already know, wouldn't that add `Promise` support without introducing the way errors are thrown?

**Comment by flackr** [comment]
Yes, I think this would result in no change to the way errors are thrown if the scrolling APIs returned an object from which you could get the finished promise. The reason the Web Animation API has this is because the Animation object (returned from element.animate) itself has a lot of useful API surface - whereas in this case we would likely just have finished on that intermediate object for now. Not to say that we couldn't do that, just pointing out that the resolved solution does align with other APIs like fetch.

I have separately thought that Element.scrolled should allow you to get the scroll promise for a scroll container that is currently scrolling, I filed #13760 for that.

**Comment by nwidynski** [comment]
Would it make sense to promisify https://developer.mozilla.org/en-US/docs/Web/API/NavigateEvent/scroll aswell? Since intercept handlers are already async I feel like that could play together nicely.

**Comment by mustaqahmed** [comment]
nwidynski@: Interesting, thanks for flagging this. Could you file a new issue at https://github.com/whatwg/html/issues? `NavigationEvent` is defined in the HTML spec.

**Comment by nwidynski** [comment]
@mustaqahmed I've got two more questions that have come up for me while working with the API:

A) Whether a scroll promise is guaranteed to resolve before a `scrollend` event. The spec only defines that there must be no more pending updates, which is identical to `scrollend` as long as the gesture has completed. Would it make sense to define an order?

B) In virtualized scroll containers with dynamic item size, a change to scroll area during scroll currently causes the scroll promise to be interrupted. This is regardless of whether or not `overflow-anchor` is set to `auto`/`none`. I'm not entirely sure if this is by design or an issue with Blinks implementation, but I feel like that shouldn't be the case.

**Comment by mustaqahmed** [comment]
> A) Whether a scroll promise is guaranteed to resolve before a `scrollend` event. The spec only defines that there must be no more pending updates, which is identical to `scrollend` as long as the gesture has completed. Would it make sense to define an order?

The order is mentioned in CSSOM View 1 [perform a scroll](https://drafts.csswg.org/cssom-view/#perform-a-scroll) algorithm. Note two important differences from "scrollend" events:
- If there is no change in scroll position, the event is not fired in Step 7.1 but the promise is still resolved in Step 7.2.
- The promise resolved in Step 7.2 is not necessarily the promise we see from JS. If multiple scrollers are affected by a `scrollIntoView` call, each of those scrollers will have a promise in Step 7.2 that remains hidden from JS, and the promise returned by the call will resolve when all those hidden promises are resolved.

> B) In virtualized scroll containers with dynamic item size, a change to scroll area during scroll currently causes the scroll promise to be interrupted. This is regardless of whether or not `overflow-anchor` is set to `auto`/`none`. I'm not entirely sure if this is by design or an issue with Blinks implementation, but I feel like that shouldn't be the case.

Please make a minimal repro and file a new issue to discuss this. I didn't see any interruption when I locally added new content to [this scroller](https://codepen.io/mustaqahmed/full/gbpmNOP) after a button click!

EDIT: added a link that I missed in the last para.
